ouroboros-network/src/Ouroboros/Network/PeerSelection/LedgerPeers.hs

```diff
@@ -0,0 +1,188 @@
+{-# LANGUAGE BangPatterns      #-}
+{-# LANGUAGE GADTs             #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Ouroboros.Network.PeerSelection.LedgerPeers (
+    DomainAddress (..),
+    IP.IP (..),
+    LedgerPeersConsensusInterface (..),
+    RelayAddress (..),
+    PoolStake,
+    runLedgerPeers,
+    TraceLedgerPeers (..),
+    pickPeers,
+    ackPoolStake,
+
+    Socket.PortNumber
+    ) where
+
+
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadTime
+import           Control.Monad.Class.MonadTimer
+import           Control.Tracer (Tracer, traceWith)
+import qualified Data.IP as IP
+import           Data.List (foldl', nub)
+import           Data.List.NonEmpty (NonEmpty (..))
+import qualified Data.List.NonEmpty as NonEmpty
+import           Data.Maybe (isNothing, fromJust)
+import qualified Data.Map.Strict as Map
+import           Data.Map.Strict (Map)
+import           Data.Ratio
+import           Data.Word
+import qualified Network.Socket as Socket
+import           System.Random
+
+import           Cardano.Slotting.Slot (SlotNo)
+import           Ouroboros.Network.PeerSelection.RootPeersDNS (DomainAddress (..))
+import           Ouroboros.Network.Subscription.Ip
+import           Ouroboros.Network.Subscription.Dns
+
+import           Text.Printf
+
+newtype LedgerPeersConsensusInterface m = LedgerPeersConsensusInterface {
+      lpGetPeers :: SlotNo -> STM m (Maybe [(PoolStake, NonEmpty RelayAddress)])
+    }
+
+data TraceLedgerPeers =
```

Comment:
Could you add haddocks, including to each constructor, e..g. what is the `Word16` in `PickedPeers`, or the `Int` in `FetchingNewlEdgerState`.